toolchain/check/impl_lookup.cpp

```diff
@@ -296,6 +298,59 @@ static auto GetWitnessIdForImpl(
       context.sem_ir(), specific_id, impl.witness_id));
 }
 
+// In the case where `facet_const_id` is a facet, see if its facet type requires
+// that `specific_interface` is implemented. If so, return the witness from the
+// facet.
+static auto FindWitnessInFacet(
+    Context& context, SemIR::LocId loc_id, SemIR::ConstantId facet_const_id,
+    const SemIR::SpecificInterface& specific_interface) -> SemIR::InstId {
+  SemIR::InstId facet_inst_id =
+      context.constant_values().GetInstId(facet_const_id);
+  // TODO: Should we convert from a FacetAccessType to its facet here?
+  SemIR::TypeId facet_type_id = context.insts().Get(facet_inst_id).type_id();
+  if (auto facet_type_inst =
+          context.types().TryGetAs<SemIR::FacetType>(facet_type_id)) {
+    auto complete_facet_type_id = RequireCompleteFacetType(
+        context, facet_type_id, loc_id, *facet_type_inst,
+        [&]() -> DiagnosticBuilder {
+          // TODO: Find test that triggers this code path.
+          CARBON_FATAL("impl lookup for with incomplete facet type");
+        });
+    if (complete_facet_type_id.has_value()) {
+      const auto& complete_facet_type =
+          context.complete_facet_types().Get(complete_facet_type_id);
+      for (auto interface : complete_facet_type.required_interfaces) {
+        if (interface == specific_interface) {
+          // TODO: Need to get the right witness when there are multiple.
```

Comment:
(I probably need to address this as part of https://github.com/carbon-language/carbon-lang/pull/5075 to avoid re-introducing crashability. Also maybe I need to adjust some language in there since the witnesses will not all be ImplWitness - but they will evaluate to one)